### PR TITLE
Rust updates and optimizations around communication performance

### PR DESCRIPTION
**Pull request by nazar-pc**
Changelog update has all major changes mentioned.

The big topic here is transition from file descriptors usage to native FFI function calls between C++ and Rust for lower latency, lower memory usage and better cross-platform support (Windows support is coming!).

The way it is implemented is the following:
* each channel construction has 2 pairs of optional parameters:
  * `channelReadFn` and `channelReadCtx` represent a function call (this is C ABI, so we can't send closures, we need static function and its context)
    * it is called by C++ when worker wants to read a message
    * with every call C++ sends over `uv_async_t` handle, so Rust (or any other language, really) can ping worker's `libuv` to try to read next message
      * because of this C++ does one initial call proactively to hand over this handle to the other side
  * `ChannelWriteCtx` and `ChannelWriteFn` also represent a similar function
    * it is called by C++ when worker wants to write a message
    * because Rust side here is thread-safe, this is immediately delivered to Rust and does all of the business logic before returning, so no need for additional notifications
* the implementation allows C++ worker to share memory with Rust using raw pointer without even copying it
* Rust though has to allocate for outgoing messages because it doesn't know when worker will be ready to read

I do not like how many arguments there is in worker instantiation now, but at the same time I don't see a huge potential for improvements either as we are limited by shared C ABI here. Open to suggestions.

### CODE REVIEW

File: worker/src/Channel/ChannelSocket.cpp
Comment (ibc) [inline worker/src/Channel/ChannelSocket.cpp]: Dot at the end of comments 😀

File: worker/src/PayloadChannel/PayloadChannelSocket.cpp
Comment (ibc) [inline worker/src/PayloadChannel/PayloadChannelSocket.cpp]: Same

File: worker/src/Channel/ChannelSocket.cpp
Comment (ibc) [inline worker/src/Channel/ChannelSocket.cpp:110]: Line break please

**Review by ibc — COMMENTED** [review]
This is awesome. Definitely I need some time before I can properly review this (Jose is gonna be out for a while for very good reasons). So just a question/concern for now:

Here comm between Rust and worker is sync meaning that if Rust layer (i.e. the Rust app) blocks or makes the CPU be super busy, that would also block and affect the worker loop. Hopefully I'm wrong given that uv_async is used in worker.

**Comment by nazar-pc** [comment]
> Here comm between Rust and worker is sync meaning that if Rust layer (i.e. the Rust app) blocks or makes the CPU be super busy, that would also block and affect the worker loop. Hopefully I'm wrong given that uv_async is used in worker.

You are correct here. When worker sends a response or notification into Rust this is definitely something everyone should keep in mind, it will steal some time from worker, but presumably very little. On the flip side the whole things is significantly more efficient, so developer now can decide whether to do memory copy and send it to another thread or process in place.

Communication from Rust to worker is async, so there is no such problem there.

File: src/Channel.ts
Comment (ibc) [inline src/Channel.ts]: Why this? it should be a string already. And BTW I'm afraid there are gonna be conflicts here now since these files are now in `node/src/` instead.

Comment (nazar-pc) [reply src/Channel.ts]: The reason is that **technically** `targetId` can be both a string (UUID) and a number (process ID), while function signature is:
```typescript
NodeJS.EventEmitter.emit(event: string | symbol, ...args: any[]): boolean
```

So using `String()` is semantically correct to use to make sure we always end up with a string there. I know it worked before, but it felt wrong.

> there are gonna be conflicts here

It is fine, I will resolve conflicts after your changes are merged.

Comment (ibc) [reply src/Channel.ts]: My changes are already merged :)

Comment (nazar-pc) [reply src/Channel.ts]: #673 still isn't :thinking:

Comment (ibc) [reply src/Channel.ts]: My fault, I was pretty sure I clicked Merge button but I didn't... done now

**Comment by nazar-pc** [comment]
Merge conflicts resolved

**Review by jmillan — APPROVED** [review]
👏 Great

**Comment by nazar-pc** [comment]
Merged latest updates from `v3` here.
@ibc do you have any thoughts on this yet?

**Comment by ibc** [comment]
I'll review tomorrow

File: worker/include/Channel/ChannelNotifier.hpp
Comment (ibc) [inline worker/include/Channel/ChannelNotifier.hpp:15]: In which case do we emit with a number instead of a string?

File: worker/include/common.hpp
Comment (ibc) [inline worker/include/common.hpp]: Please dot after all comments here.

**Review by ibc — CHANGES_REQUESTED** [review]

File: worker/src/Channel/ChannelSocket.cpp
Comment (ibc) [inline worker/src/Channel/ChannelSocket.cpp]:
```suggestion
	/* Static methods for UV callbacks. */
```

File: worker/src/Worker.cpp
Comment (ibc) [inline worker/src/Worker.cpp:40]: I see here.

**Review by ibc — COMMENTED** [review]
Just a cosmetic change. Code looks great.

**Review by ibc — APPROVED** [review]
PR 666 approved (once CI doesn't complain)

**Comment by ibc** [comment]
@nazar-pc feel free to merge :)

**Pull request merged**
